funcframework/events_test.go

```diff
@@ -134,45 +134,142 @@ func TestGetBackgroundEvent(t *testing.T) {
+				Resource: &metadata.Resource{
+					RawPath: "projects/sample-project/topics/gcf-test",
+				},
+			},
+			wantData: map[string]interface{}{
+				"data": "VGhpcyBpcyBhIHNhbXBsZSBtZXNzYWdl",
+			},
+		},
+		{
+			name: "data and invalid embedded context event no error",
+			body: []byte(`{
+  "data": {
+    "data": "VGhpcyBpcyBhIHNhbXBsZSBtZXNzYWdl"
+  }
+}`),
+		},
+		{
+			name: "raw pubsub event",
+			body: []byte(`{
+				"subscription": "projects/FOO/subscriptions/BAR_SUB",
+				"message": {
+					"data": "eyJmb28iOiJiYXIifQ==",
+					"messageId": "1",
+					"attributes": {
+						"test": "123"
+					}
+				}
+			}`),
+			url: "projects/sample-project/topics/gcf-test",
+			wantMetadata: &metadata.Metadata{
+				EventID:   "1",
+				EventType: "google.pubsub.topic.publish",
+				Resource: &metadata.Resource{
+					Name:    "projects/sample-project/topics/gcf-test",
+					Type:    "type.googleapis.com/google.pubusb.v1.PubsubMessage",
+					Service: "pubsub.googleapis.com",
+				},
+			},
+			wantData: map[string]interface{}{
+				"@type": "type.googleapis.com/google.pubusb.v1.PubsubMessage",
+				"data":  []byte(`{"foo":"bar"}`),
+				"attributes": map[string]string{
+					"test": "123",
+				},
+			},
+		},
+		{
+			name: "raw pubsub event missing url",
+			body: []byte(`{
+				"subscription": "projects/FOO/subscriptions/BAR_SUB",
+				"message": {
+					"data": "eyJmb28iOiJiYXIifQ==",
+					"messageId": "1",
+					"attributes": {
+						"test": "123"
+					}
+				}
+			}`),
+			wantMetadata: &metadata.Metadata{
+				EventID:   "1",
+				EventType: "google.pubsub.topic.publish",
+				Resource: &metadata.Resource{
+					Type:    "type.googleapis.com/google.pubusb.v1.PubsubMessage",
+					Service: "pubsub.googleapis.com",
+				},
+			},
+			wantData: map[string]interface{}{
+				"@type": "type.googleapis.com/google.pubusb.v1.PubsubMessage",
+				"data":  []byte(`{"foo":"bar"}`),
+				"attributes": map[string]string{
+					"test": "123",
+				},
+			},
+		},
 	}
 
 	for _, tc := range tcs {
 		t.Run(tc.name, func(t *testing.T) {
-			md, d, err := getBackgroundEvent(tc.input)
+			md, d, err := getBackgroundEvent(tc.body, tc.url)
 			if tc.hasErr && err == nil {
 				t.Errorf("expected error but got nil")
 			}
 			if !tc.hasErr && err != nil {
 				t.Errorf("expected no error, got error: %v", err)
 			}
 
-			if !cmp.Equal(md, tc.metadata) {
-				t.Errorf("incorrect metadata, got %+v, want %+v", md, tc.metadata)
+			// If timestamp is not being tested in this test case, skip comparing the field
+			// since some timestamps are auto-populated with time.Now()
+			diffOpts := []cmp.Option{}
+			if tc.wantMetadata != nil && tc.wantMetadata.Timestamp.IsZero() {
+				diffOpts = append(diffOpts, cmpopts.IgnoreFields(metadata.Metadata{}, "Timestamp"))
+			}
+
+			if diff := cmp.Diff(tc.wantMetadata, md, diffOpts...); diff != "" {
+				t.Errorf("MakeGatewayInfo() mismatch (-want +got):\n%s", diff)
```

Comment:
Is `MakeGatewayInfo` a typo?

Reply:
Oh, yeah. It's the example from the cmp.Diff docs :)